### PR TITLE
#1746 - Send emails in batches

### LINKED ISSUE #1746: Send emails in batches
**Is your feature request related to a problem? Please describe.**
In platform, we send emails to all members in one go and it works fine but when the site has a lot of users like in thousands or millions, it won't work and it will break there.

**Describe the solution you'd like**
We should send emails in batch/parts and send emails for all components.

**Describe alternatives you've considered**
No alt.

### Scope
We will create a DB table to list down the email queue. When an email is requested to send, it will be added to the queue.
A cron job will run every 2 minutes and will send max 20 emails every time and will delete the items which is sent.

### Related Platform API issue
https://github.com/buddyboss/buddyboss-platform-api/issues/315

### Related Platform PRO issue
https://github.com/buddyboss/buddyboss-platform-pro/issues/190

### PR DESCRIPTION
**Pull request by sadrul**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:
 
A DB table is created to list down the email queue. When an email is requested to send, it will be added to the queue. Background process batch will be created taking 20 items from there and will execute in the background. The Db table data will be cleared as soon as it is processed.

### Changes in Logic
The batch email will be implemented only where more than 20 emails need to process and a filter will be provided to change this number. Otherwise, it will be processed in the default way

Fixes #1746  .

### How to test the changes in this Pull Request:
It can be tested using any action which sends an email like friend request sent, private message send, group message send, add topic/reply. Below is example steps for sending a group message

1. Go to a group page
2. Switch to group admin
3. Send message > Send group message

### Proof Screenshots or Video
https://www.loom.com/share/f84b7f73e38e43e5a80f6d13d88543bb

### Blocker
This task is dependent on this PR https://github.com/buddyboss/buddyboss-platform/pull/3061

### How to use

Global function `bp_email_queue()` which instantiates the class

Add a record in the email queue table
`bp_email_queue()->add_record( $email_type, $receiver_user_id, $args );`

Call email background process.
`bp_email_queue()->bb_email_background_process();`

So, we need to replace the code where `bp_send_email` is used like this - 
```
		if ( function_exists( 'bp_email_queue' ) ) {
			bp_email_queue()->add_record( $email_type, $receiver_user_id, $args );
			// call email background process.
			bp_email_queue()->bb_email_background_process();
		} else {
			bp_send_email( $email_type, $receiver_user_id, $args );
		}
```

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Related Platform API PR
https://github.com/buddyboss/buddyboss-platform-api/pull/316

### Related Platform PRO PR
https://github.com/buddyboss/buddyboss-platform-pro/pull/191

### Implemented areas

**Platform**

Forums
- Sends notification emails for new replies to subscribed topics
- Sends notification emails for new topics to subscribed forums

Groups
- Notify all group members when a group is updated.
- Email message recipients to alert them of a new unread group message.

Messages
- Email message recipients to alert them of a new unread private message.

Profile
- Member submit email invite.

**API**
- Create an Invites/Send Invites

**PRO**
- Send meeting notifications for group
- Send webinar notifications for group.

### CODE REVIEW

File: src/bp-core/classes/class-bp-email-queue.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	 */
+	class BP_Email_Queue {
+
+		/**
+		 * Constructor.
+		 *
+		 * @since BuddyBoss 1.7.6
+		 */
+		function __construct() {
+			$this->create_db_table();
```
Comment (KartikSuthar) [inline src/bp-core/classes/class-bp-email-queue.php:26]: remove this from here

File: src/bp-core/classes/class-bp-email-queue.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+if ( ! class_exists( 'BP_Email_Queue' ) ) :
+
+	/**
+	 * Load Email queue class
+	 *
+	 * @since BuddyBoss 1.7.6
+	 */
+	class BP_Email_Queue {
+
```
Comment (KartikSuthar) [inline src/bp-core/classes/class-bp-email-queue.php:19]:
```
 public static function instance() {
        if ( is_null( self::$_instance ) )
			self::$_instance = new self();

		return self::$_instance;
    }

```

Comment (KartikSuthar) [reply src/bp-core/classes/class-bp-email-queue.php:19]: `private static $_instance = null;`

File: src/bp-core/classes/class-bp-email-queue.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		 * Set up Cron.
+		 *
+		 * @since BuddyBoss 1.7.6
+		 */
+		function setup_cron() {
+			if ( ! wp_next_scheduled( 'bp_email_queue_cron_hook' ) ) {
+				wp_schedule_event( time(), 'bb_schedule_1min', 'bp_email_queue_cron_hook' );
+			}
+
+			add_action( 'bp_email_queue_cron_hook', array( $this, 'email_queue_cron_cb' ) );
```
Comment (KartikSuthar) [inline src/bp-core/classes/class-bp-email-queue.php:40]: replace with background job here

File: src/bp-core/classes/class-bp-email-queue.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+			return $wpdb->query( $wpdb->prepare( "DELETE FROM {$wpdb->prefix}bb_email_queue WHERE id = %d", $id ) );
+		}
+
+		/**
+		 * Email queue get record
+		 *
+		 * @since BuddyBoss 1.7.6
+		 */
+		public function get_records() {
```
Comment (KartikSuthar) [inline src/bp-core/classes/class-bp-email-queue.php:70]:
```suggestion
		public function get_records( $limit = 20, $order_column = 'id', $orderby = 'ASC' ) {
```

File: src/bp-core/classes/class-bp-email-queue.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+			if ( ! function_exists( 'dbDelta' ) ) {
+				require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+			}
+
+			$sql = "CREATE TABLE {$wpdb->prefix}bb_email_queue (
+	            id bigint(20) NOT NULL AUTO_INCREMENT,
+	            email_type varchar(200) NOT NULL,
+	            recipient_id bigint(20) NOT NULL DEFAULT 0,
+	            arguments mediumtext DEFAULT NULL,
```
Comment (KartikSuthar) [inline src/bp-core/classes/class-bp-email-queue.php:114]:
```suggestion
	            arguments longtext DEFAULT NULL,
```

File: src/bp-core/classes/class-bp-email-queue.php [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	            date_created datetime DEFAULT '0000-00-00 00:00:00' NOT NULL,
+	            PRIMARY KEY  (id)
+	        ) $charset_collate;";
+
+			dbDelta( $sql );
+		}
+	}
+
```
Comment (KartikSuthar) [inline src/bp-core/classes/class-bp-email-queue.php:122]:
```
function bp_email_queue() {
	if ( class_exists( 'BP_Email_Queue' ) ) {
		global $bp_email_queue;
		$bp_email_queue = BP_Email_Queue::instance();

		return $bp_email_queue;
	}
}
```

File: src/bp-core/bp-core-email-queue.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ */
+function bp_email_queue_init() {
+	bp_email_queue();
+}
+
+add_action( 'bp_init', 'bp_email_queue_init' );
+
+
+
+
```
Comment (KartikSuthar) [inline src/bp-core/bp-core-email-queue.php:41]: remove this file its not needed.

File: src/class-buddypress.php [hunk 8]
```diff
@@ -554,6 +554,7 @@ private function includes() {
 		require $this->plugin_dir . 'bp-core/bp-core-emails.php';
 		require $this->plugin_dir . 'bp-core/bp-core-wp-emails.php';
 		require $this->plugin_dir . 'bp-core/bp-core-learndash-emails.php';
+		require $this->plugin_dir . 'bp-core/bp-core-email-queue.php';
```
Comment (KartikSuthar) [inline src/class-buddypress.php:557]: remove the file include here

**Review by KartikSuthar — CHANGES_REQUESTED** [review]

File: src/bp-core/classes/class-bp-email-queue.php [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+		if ( isset( $get_records ) && ! empty( $get_records ) ) {
+			foreach ( $get_records as $single ) {
+				$item_id    = ! empty( $single->id ) ? $single->id : 0;
+				$email_type = ! empty( $single->email_type ) ? $single->email_type : '';
+				$to         = ! empty( $single->recipient ) ? maybe_unserialize( $single->recipient ) : 0;
+				$args       = ! empty( $single->arguments ) ? maybe_unserialize( $single->arguments ) : array();
+				//if ( bp_send_email( $email_type, $to, $args ) ) {
+					bp_send_email( $email_type, $to, $args );
+					$this->delete_record( $item_id );
+				//}
```
Comment (chetansatasiya) [inline src/bp-core/classes/class-bp-email-queue.php:127]: @sadrul We need to remove the record only if the email fired successfully. So need to keep the condition as it is.

File: src/bp-groups/bp-groups-notifications.php [hunk 10]
```diff
@@ -295,7 +319,11 @@ function groups_notification_promoted_member( $user_id = 0, $group_id = 0 ) {
 			'unsubscribe' => esc_url( bp_email_get_unsubscribe_link( $unsubscribe_args ) ),
 		),
 	);
-	bp_send_email( 'groups-member-promoted', (int) $user_id, $args );
+	bp_email_queue()->add_record( 'groups-member-promoted', (int) $user_id, $args );
+	if ( function_exists( 'bp_email_queue' ) ) {
+		// call email background process.
+		bp_email_queue()->bb_email_background_process();
+	}
```
Comment (chetansatasiya) [inline src/bp-groups/bp-groups-notifications.php:326]:
```suggestion
	if ( function_exists( 'bp_email_queue' ) ) {
	        bp_email_queue()->add_record( 'groups-member-promoted', (int) $user_id, $args );
		// call email background process.
		bp_email_queue()->bb_email_background_process();
	} else {
	       bp_send_email( 'groups-member-promoted', (int) $user_id, $args );
	}
```

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
Need to check class exists in many places

File: src/bp-core/bp-core-update.php [hunk 11]
```diff
@@ -337,6 +337,10 @@ function bp_version_updater() {
 		if ( $raw_db_version < 17401 ) {
 			bb_update_to_1_7_5();
 		}
+
+		if ( $raw_db_version < 17501 ) {
```
Comment (KartikSuthar) [inline src/bp-core/bp-core-update.php:341]:
```suggestion
		if ( $raw_db_version < 17701 ) {
```

File: src/class-buddypress.php [hunk 12]
```diff
@@ -330,7 +330,7 @@ private function setup_globals() {
 
 		/** Versions */
 		$this->version    = defined( 'BP_PLATFORM_VERSION' ) ? BP_PLATFORM_VERSION : ( defined( 'BP_VERSION' ) ? BP_VERSION : '1.0.0' );
-		$this->db_version = 17401;
+		$this->db_version = 17501;
```
Comment (KartikSuthar) [inline src/class-buddypress.php:333]:
```suggestion
		$this->db_version = 17701;
```

**Review by KartikSuthar — CHANGES_REQUESTED** [review]

File: src/bp-invites/actions/invites.php [hunk 13]
```diff
@@ -141,7 +141,11 @@ function bp_member_invite_submit() {
 		 */
 		add_filter( 'bp_email_get_salutation', '__return_false' );
 		// Send invitation email.
-		bp_send_email( 'invites-member-invite', $email, $args );
+		bp_email_queue()->add_record( 'invites-member-invite', $email, $args );
+		if ( function_exists( 'bp_email_queue' ) ) {
+			// call email background process.
+			bp_email_queue()->bb_email_background_process();
+		}
```
Comment (KartikSuthar) [inline src/bp-invites/actions/invites.php:148]:
```suggestion
              if ( function_exists( 'bp_email_queue' ) ) {
			bp_email_queue()->add_record( 'invites-member-invite', $email, $args );
			// call email background process.
			bp_email_queue()->bb_email_background_process();
		} else {
			bp_send_email( 'invites-member-invite', $email, $args );
		}
```

File: src/bp-members/bp-members-functions.php [hunk 14]
```diff
@@ -2311,7 +2311,12 @@ function bp_core_signup_send_validation_email( $user_id, $user_email, $key, $use
 		$to = array( array( $user_email => $user_login ) );
 	}
 
-	bp_send_email( 'core-user-registration', $to, $args );
+	bp_email_queue()->add_record( 'core-user-registration', $to, $args );
+
+	if ( function_exists( 'bp_email_queue' ) ) {
+		// call email background process.
+		bp_email_queue()->bb_email_background_process();
+	}
```
Comment (KartikSuthar) [inline src/bp-members/bp-members-functions.php:2319]:
```suggestion
	        if ( function_exists( 'bp_email_queue' ) ) {
			bp_email_queue()->add_record( 'core-user-registration', $to, $args );
			// call email background process.
			bp_email_queue()->bb_email_background_process();
		} else {
			bp_send_email( 'core-user-registration', $to, $args );
		}
```

**Review by KartikSuthar — APPROVED** [review]

**Comment by sadrul** [comment]
@sumitbuddyboss @mohit-buddyboss merged into dev and qa-ongoing and ready for testing

**Comment by sadrul** [comment]
This task is bumped for release 1.7.7 because it was not delivered in keeping enough time for QA in hand. The task has taken time because there are a couple of areas that need to consider implementing this and had to test a lot to make sure working fine for each module and for a varied range of users. Also, we had to restructure the code couple of times to achieve this correctly, so there was a couple of back and forth and discussion. There are the main reasons for delaying the delivery of this task.

File: src/bp-core/bp-core-update.php [hunk 15]
```diff
@@ -337,6 +337,14 @@ function bp_version_updater() {
 		if ( $raw_db_version < 17401 ) {
 			bb_update_to_1_7_5();
 		}
+
+		if ( $raw_db_version < 17701 ) {
+			bb_update_to_1_7_7();
+		}
```
Comment (KartikSuthar) [inline src/bp-core/bp-core-update.php:343]: remove this function not exists.

File: src/bp-core/bp-core-update.php [hunk 16]
```diff
@@ -715,6 +723,18 @@ function bb_update_to_1_7_5() {
 
 }
 
+/**
+ * Function to update data
+ *
+ * @since BuddyBoss 1.7.8
+ */
+function bb_update_to_1_7_8() {
+	if ( function_exists( 'bb_is_email_queue' ) && bb_is_email_queue() ) {
```
Comment (KartikSuthar) [inline src/bp-core/bp-core-update.php:732]:
```suggestion
	if ( function_exists( 'bb_email_queue' )  ) {
```

**Review by KartikSuthar — CHANGES_REQUESTED** [review]

File: src/class-buddypress.php [hunk 17]
```diff
@@ -330,7 +330,7 @@ private function setup_globals() {
 
 		/** Versions */
 		$this->version    = defined( 'BP_PLATFORM_VERSION' ) ? BP_PLATFORM_VERSION : ( defined( 'BP_VERSION' ) ? BP_VERSION : '1.0.0' );
-		$this->db_version = 17401;
+		$this->db_version = 17701;
```
Comment (KartikSuthar) [inline src/class-buddypress.php:333]:
```suggestion
		$this->db_version = 17801;
```

**Review by KartikSuthar — APPROVED** [review]

**Comment by sadrul** [comment]
@sumitbuddyboss it will not work in group > send message > Send private message https://i.imgur.com/rcTq0B8.png because there is some other issue (maximum execution timeout) and its been taken care in another PR https://github.com/buddyboss/buddyboss-platform/pull/2911

**Review by KartikSuthar — DISMISSED** [review]

**Comment by sumitbuddyboss** [comment]
Hello @sadrul 

**Issue:**
moderation >  If the user gets suspend and report any activity then why getting 3 batches in the database table.

**Issue:**
https://www.loom.com/share/e77339974c1a41a69d30b6e87d79497d

Thank you

**Comment by sadrul** [comment]
@sumitbuddyboss It has been fixed and merged already https://www.loom.com/share/b7301d725ecf46eabe34db5b5b5b2540

**Comment by sadrul** [comment]
We are bumping this task as we going to change the logic in the implementation. The batch email will be implemented only where more than 20 emails need to process and a filter will be provided to change this number.

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
Need changes

**Review by chetansatasiya — APPROVED** [review]

**Comment by sadrul** [comment]
@mohit-buddyboss @sumitbuddyboss Merged into dev and qa-ongoing and ready for testing

**Comment by sumitbuddyboss** [comment]
Hello @sadrul 

I have checked in qa-ongoing branch and now as per new logic Email batch is working as expected, Please check my below comments and video proof.

**Forums**
Sends notification emails for new replies to subscribed topics
Sends notification emails for new topics to subscribed forums
https://www.loom.com/share/3522a17bd195448eb1e739d66ece23d2

**Groups**
Notify all group members when a group is updated.
Email message recipients to alert them of a new unread group message.

https://www.loom.com/share/af4ef6fe51a14907ba221615452888da
https://www.loom.com/share/7ba7383fff0e4a368b723f1a7289b7d6

**Messages**
Email message recipients to alert them of a new unread private message.
https://www.loom.com/share/50cafed2845a4975bb8c728a21c13237
https://www.loom.com/share/5cf2fbafe2724e06a66df7a4f7645070

**Profile**
Member submit email invite.
Right now only we can invite 20 member so email not send in batch but our batch email code is there so if in future we increase limit then batch email is work.
https://www.loom.com/share/8759a160e7a946fab9fbf13525e6114c


Ready for release 🚀 

Thank you

**Comment by mohit-buddyboss** [comment]
@sadrul 

Tested on qa-approved branch of the platform.
This PR functionality is working as expected and ready for production.

cc @sourabhmatolia @chetansatasiya @KartikSuthar

**Comment by sadrul** [comment]
@sumitbuddyboss as we discussed via a call that the batch email background processes never execute due to the moderation module issue (suspend background processes exist in db and try to process, however they never been completed) which has been fixed in this pr https://github.com/buddyboss/buddyboss-platform/pull/3061 otherwise if we remove those suspend background processes, then batch email working fine

**Comment by JetalDobariya** [comment]
Hello @sadrul 

I did verify all of above TCs in the qa-approved branch, now all the TCs are passed and batch email functionalities are working as expected.
here , I attached few Screenshots for the proofs.
Group update :
Database : https://prnt.sc/1y9g25r
Email : https://prnt.sc/1y9g5bu

Group msg :
Database : https://prnt.sc/1y9gibc
Email : https://prnt.sc/1y9glyc

Forum : 
Databse : https://prnt.sc/1ya0j0e
Email : https://prnt.sc/1ya0ki5

Forum Reply
Database : https://prnt.sc/1ya1k4g
Email : https://prnt.sc/1ya1lo1

Zoom 
Databse : https://prnt.sc/1ya1uhc

Thanks!
CC : @sourabhmatolia @KartikSuthar @sumitbuddyboss

**Comment by KartikSuthar** [comment]
@sourabhmatolia As discussed on call we have an issue with the moderation background process(https://github.com/buddyboss/buddyboss-platform/issues/3153). 
So, as discussed we are going to create a new background process for the email one.

CC: @chetansatasiya @firozwp

**Review by chetansatasiya — DISMISSED** [review]

**Comment by KartikSuthar** [comment]
@sumitbuddyboss @JetalDobariya 

We need to test it again as a registered new background job for the email one.
http://prntscr.com/1z48yu4

We need to test the background job is correctly registered and email sending is working correctly or not.

**Comment by JetalDobariya** [comment]
Hello @KartikSuthar 

I did verify again all the Batch email which i tested earlier and all the emails are working fine.

Thanks!
CC : @sourabhmatolia @sumitbuddyboss

**Comment by azanoni** [comment]
Hi, I think that you introduce by mistake another BIG issue on 1.8.1, we updated from 1.8.0 to 1.8.2 on my client website the 1st of December, and in that time we have our wp_options table with a size of **31MB** and now we have wp_options table with a size of **60GIGABYTES!!!**, we suffer a lots of crashes/frozen DB because this, please dont attach in WP_OPTIONS without delete later this kind of rows:

SELECT count(*) FROM ebdbprd.wp_options WHERE option_name like "wp_1_bb_email%" returns 67913 rows 

image

image

Example of each json that you store in wp_options table

`a:238:{i:0;a:2:{s:8:"callback";a:2:{i:0;O:14:"BP_Email_Queue":0:{}i:1;s:22:"bb_email_queue_cron_cb";}s:4:"args";a:1:{i:0;a:1:{i:0;a:6:{s:2:"id";s:4:"1561";s:10:"email_type";s:28:"zoom-scheduled-meeting-email";s:9:"recipient";s:1024:"O:7:"WP_User":8:{s:4:"data";O:8:"stdClass":12:{s:2:"ID";s:4:"3958";s:10:"user_login";s:8:"XXXX";s:9:"user_pass";s:34:"XXXXXXXXXXX1";s:13:"user_nicename";s:8:"XXXXX";s:10:"user_email";s:21:"XXXX@gmail.com";s:8:"user_url";s:0:"";s:15:"user_registered";s:19:"2017-03-18 21:14:01";s:19:"user_activation_key";s:0:"";s:11:"user_status";s:1:"0";s:12:"display_name";s:17:"xxxx( xxxxxx)";s:4:"spam";s:1:"0";s:7:"deleted";s:1:"0";}s:2:"ID";i:3958;s:4:"caps";a:2:{s:10:"subscriber";b:1;s:15:"bbp_participant";b:1;}s:7:"cap_key";s:15:"wp_capabilities";s:5:"roles";a:2:{i:0;s:10:"subscriber";i:1;s:15:"bbp_participant";}s:7:"allcaps";a:13:{s:4:"read";b:1;s:7:"level_0";b:1;s:16:"propanel_widgets";b:0;s:8:"spectate";b:1;s:11:"participate";b:1;s:19:"read_private_forums";b:1;s:14:"publish_topics";b:1;s:11:"edit_topics";b:1;s:15:"publish_replies";b:1;s:12:"edit_replies";b:1;s:17:"assign_topic_tags";b:1;s:10:"subscriber";b:1;s:15:"bbp_participant";b:1;}s:6:"filter";N;s:16:"`

**Comment by chetansatasiya** [comment]
> Hi, I think that you introduce by mistake another BIG issue on 1.8.1, we updated from 1.8.0 to 1.8.2 on my client website the 1st of December, and in that time we have our wp_options table with a size of **31MB** and now we have wp_options table with a size of **60GIGABYTES!!!**, we suffer a lots of crashes/frozen DB because this, please dont attach in WP_OPTIONS without delete later this kind of rows:
> 
> SELECT count(*) FROM ebdbprd.wp_options WHERE option_name like "wp_1_bb_email%" returns 67913 rows
> 
> image
> 
> image
> 
> Example of each json that you store in wp_options table
> 
> `a:238:{i:0;a:2:{s:8:"callback";a:2:{i:0;O:14:"BP_Email_Queue":0:{}i:1;s:22:"bb_email_queue_cron_cb";}s:4:"args";a:1:{i:0;a:1:{i:0;a:6:{s:2:"id";s:4:"1561";s:10:"email_type";s:28:"zoom-scheduled-meeting-email";s:9:"recipient";s:1024:"O:7:"WP_User":8:{s:4:"data";O:8:"stdClass":12:{s:2:"ID";s:4:"3958";s:10:"user_login";s:8:"XXXX";s:9:"user_pass";s:34:"XXXXXXXXXXX1";s:13:"user_nicename";s:8:"XXXXX";s:10:"user_email";s:21:"XXXX@gmail.com";s:8:"user_url";s:0:"";s:15:"user_registered";s:19:"2017-03-18 21:14:01";s:19:"user_activation_key";s:0:"";s:11:"user_status";s:1:"0";s:12:"display_name";s:17:"xxxx( xxxxxx)";s:4:"spam";s:1:"0";s:7:"deleted";s:1:"0";}s:2:"ID";i:3958;s:4:"caps";a:2:{s:10:"subscriber";b:1;s:15:"bbp_participant";b:1;}s:7:"cap_key";s:15:"wp_capabilities";s:5:"roles";a:2:{i:0;s:10:"subscriber";i:1;s:15:"bbp_participant";}s:7:"allcaps";a:13:{s:4:"read";b:1;s:7:"level_0";b:1;s:16:"propanel_widgets";b:0;s:8:"spectate";b:1;s:11:"participate";b:1;s:19:"read_private_forums";b:1;s:14:"publish_topics";b:1;s:11:"edit_topics";b:1;s:15:"publish_replies";b:1;s:12:"edit_replies";b:1;s:17:"assign_topic_tags";b:1;s:10:"subscriber";b:1;s:15:"bbp_participant";b:1;}s:6:"filter";N;s:16:"`
> 
> Could you please tell us what we could do until you fix this in your core? we are increasing the DB table per 10gigabytes each week
> 
> EDIT2: You need to find a way to deleted those notifications, also improve the way to do that when the groups are having 1000 users.

@azanoni Could you please create a support ticket with us via your BuddyBoss account and send us your site's credentials so we can take look into that in priority as no one has reported this kind of issue to us?

**Comment by chetansatasiya** [comment]
@azanoni Also, in the meantime you can add this code `add_filter( 'bb_is_email_queue', '__return_false' );` to your child theme `functions.php` file to stop creating a cron and fire the email instantly without creating a background job.

**Pull request merged**